From 970c5bde9b8abcbc7ad5f7189313238fe727e538 Mon Sep 17 00:00:00 2001
From: Stephan Gerhold <stephan.gerhold@linaro.org>
Date: Fri, 25 Jul 2025 13:07:37 +0200
Subject: [PATCH 079/150] soc: qcom: smp2p: Kick after requesting interrupt

There is a small chance that an already running remoteproc does not take
kicking it lightly and kicks us back immediately after. If this happens
during probe() there is an even smaller chance that we might miss the
incoming interrupt, because it is not registered yet.

Avoid this race condition by kicking the outgoing edge only after we
started listening for incoming interrupts.

Fixes: 50e99641413e ("soc: qcom: smp2p: Qualcomm Shared Memory Point to Point")
Signed-off-by: Stephan Gerhold <stephan.gerhold@linaro.org>
Signed-off-by: Abel Vesa <abel.vesa@oss.qualcomm.com>
---
 drivers/soc/qcom/smp2p.c | 6 +++---
 1 file changed, 3 insertions(+), 3 deletions(-)

diff --git a/drivers/soc/qcom/smp2p.c b/drivers/soc/qcom/smp2p.c
index cb515c234..1ea4d35c6 100644
--- a/drivers/soc/qcom/smp2p.c
+++ b/drivers/soc/qcom/smp2p.c
@@ -618,9 +618,6 @@ static int qcom_smp2p_probe(struct platform_device *pdev)
 		}
 	}
 
-	/* Kick the outgoing edge after allocating entries */
-	qcom_smp2p_kick(smp2p);
-
 	ret = devm_request_threaded_irq(&pdev->dev, irq,
 					NULL, qcom_smp2p_intr,
 					IRQF_ONESHOT,
@@ -630,6 +627,9 @@ static int qcom_smp2p_probe(struct platform_device *pdev)
 		goto unwind_interfaces;
 	}
 
+	/* Kick the outgoing edge after allocating entries */
+	qcom_smp2p_kick(smp2p);
+
 	/*
 	 * Treat smp2p interrupt as wakeup source, but keep it disabled
 	 * by default. User space can decide enabling it depending on its
-- 
2.52.0

